test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx

```diff
@@ -176,8 +185,211 @@ BOOST_AUTO_TEST_CASE(test_export_coreml) {
+      throw std::ios_base::failure(ss.str());
+    }
+    TS_ASSERT_EQUALS(file_size, file.tellg());
+    file_size = file.tellg();
+    // reset the cursor
+    file.seekg(0, std::ios::beg);
+
+    std::vector<char> model_spec_loaded(file_size);
+    file.read(model_spec_loaded.data(), file_size);
+    file.close();
+
+    // compare
+    TS_ASSERT_EQUALS(std::strncmp(model_spec_loaded.data(),
+                                  model_spec_src.data(), file_size),
+                     0);
+
+    // clean myself
+    if (std::remove(my_file)) {
+      std::stringstream ss;
+      ss << "fail to remove file:" << my_file << std::endl;
+      std::perror(ss.str().c_str());
+    }
+    if (std::remove(my_file_loaded)) {
+      std::stringstream ss;
+      ss << "fail to remove file:" << my_file_loaded << std::endl;
+      std::perror(ss.str().c_str());
+    }
+
+    // compare weights in memory
+    auto original_spec = dc.get_model_spec_copy();
+    TS_ASSERT(original_spec != nullptr);
+    auto original_view = original_spec->export_params_view();
+
+    auto loaded_spec = dc_other.get_model_spec_copy();
+    TS_ASSERT(loaded_spec != nullptr);
+    auto loaded_view = loaded_spec->export_params_view();
+
+    TS_ASSERT(original_view.size() > 1);
+    TS_ASSERT_EQUALS(original_view.size(), loaded_view.size());
+
+    for (const auto& entry : original_view) {
+      // std::cout << entry.first << std::endl;
+      auto& loaded_weights = loaded_view.at(entry.first);
+      auto& original_weights = entry.second;
+      TS_ASSERT(loaded_weights.size() > 0);
+      TS_ASSERT_EQUALS(original_weights.size(), loaded_weights.size());
+      TS_ASSERT_EQUALS(
+       std::memcmp(loaded_weights.data(), original_weights.data(),
+                       loaded_weights.size()),
+          0);
+    }
+  };
+
+  // states
+  constexpr unsigned int num_classes = 10;
+  const std::string target = "target";
+  const std::vector<std::string> features = {"0", "1"};
+
+  // random init; avoid segfault
+
+  drawing_classifier_mock dummy;
+  dummy.add_or_update_state(
+      {{"target", target},
+       {"num_classes", num_classes},
+       {"random_seed", 1},
+       {"features", flex_list(features.begin(), features.end())}});
+
+  // model spec should be different since all weights are random
+  // generated
+  auto spec1 = dummy.get_model_spec();
+  TS_ASSERT(spec1 != nullptr);
+
+  dummy.add_or_update_state({{"random_seed", 2}});
+  auto spec2 = dummy.get_model_spec();
+  TS_ASSERT(spec2 != nullptr);
+
+  auto view1 = spec1->export_params_view();
+  auto view2 = spec2->export_params_view();
+
+  std::vector<int> is_same;
+
+  for (const auto& entry : view1) {
+    auto& weights2 = view2.at(entry.first);
+    auto& weights1 = entry.second;
+    TS_ASSERT(weights1.size() > 0);
+    TS_ASSERT_EQUALS(weights1.size(), weights2.size());
+    is_same.push_back(
+        std::memcmp(weights1.data(), weights2.data(), weights1.size()));
+  }
```

Comment:
Why is there a layer-by-layer comparison here and in the implementation of `load_save_compare`? All of this feels unnecessarily complex

Reply:
because I want to make sure the initial`nn_spec` of 2 mock instances are different.